### ISSUE TITLE
Spike: Research new certificate generation process/libraries

### ISSUE THREAD

**Issue by rsgonzal** [Iridescent-CM/technovation-app#2953]
Goal of ticket: Research what libraries are available for certificate generation per the requirements listed below. Share information/recommendation with the team so a final decision can be made.

**Comment by dboyer** [comment]
I believe this references the work partially done in https://github.com/Iridescent-CM/technovation-app/issues/2952 to make certificate generation ad hoc rather than in bulk

**Comment by dboyer** [comment]
The current certificate generation process takes a long time, often misses certificates (#2916), and cannot print Cyrillic characters (#2591). We should select a new library that:

- Allows non-Roman characters
- Supports ad-hoc certificate generation
- Enables certificates to be available for past seasons through the user's dashboard. This is part of moving toward a platform that has the concept of seasonality.

**Comment by shaunxp20** [comment]
## Best New Option (IMHO)

### HexaPDF

https://hexapdf.gettalong.org/index.html

From [The Ruby Toolbox](https://www.ruby-toolbox.com/projects/hexapdf):
> A long-lived project that still receives updates

From the README:
> (it's a tool for) manipulating existing PDF files

> This tool is intended to be a replacement for tools like pdftk and the various Poppler-based tools like pdfinfo, pdfimages, 

Just FYI, `pdftk` is what we use via the `pdf-forms` gem.

I think HexaPDF is the best option for us to explore more/spike on.

## Other Options

### CombinePDF

https://github.com/boazsegev/combine_pdf

Looks like it will do what we want, but there could be some potential issues w/ filling in PDF forms:
https://github.com/boazsegev/combine_pdf#known-limitations

### Wicked PDF

https://github.com/mileszs/wicked_pdf

Similar to to our library that converts HTML to PDFs in that it uses `wkhtmltopdf` under the hood. Has a downside in that `wkhtmltopdf` is in archive mode/no longer being maintained.

### Origami

https://github.com/gdelugre/origami

>Origami is a framework written in pure Ruby to manipulate PDF files.

> It offers the possibility to parse the PDF contents, modify and save the PDF structure, as well as creating new documents.

Looks like it will do what we want but, it hasn't been updated in over three years.

## Paid Options

### Apryse/PDFTron

Looks very simple to fill in PDF forms:
https://docs.apryse.com/documentation/ruby/guides/features/forms/fill-fields/

#### Pricing

Couldn't find any info online regarding pricing, but here's the pricing page:
https://apryse.com/pricing

### PDFlib

https://www.pdflib.com/products/pdflib-family/whats-new-in-pdflib-10/

> PDFlib is a well-established developer component for generating and manipulating PDF documents. PDFlib’s main targets are dynamic PDF creation on a Web server and PDF creation in any kind of server, desktop or mobile application. You can use PDFlib to dynamically create PDF documents from database contents, similar to dynamic Web pages. Since 1997 PDFlib has proven itself in a wide range of use cases and development environments. Application programmers use PDFlib to create PDF output with text, images, graphics and interactive elements.

There's a guide/docs for Ruby too.

#### Pricing
https://www.pdflib.com/buy/online-shop/pdflib-pdflib-pdi-pps/

---

## Other Notes/Comments

Regarding our current library (`pdf-forms`), from the README:
> ...your input PDF lacks the proper font for your kind of characters. Re-create it and embed any necessary fonts.

https://github.com/jkraemer/pdf-forms#non-ascii-characters-utf8-etc-are-not-displayed-in-the-filled-out-pdf

It's possible the fonts we were using didn't support UTF-8 characters, might be something for us to look into.

I also made a comment about using our current "HTML to PDF" gem ('pdfkit`) to fill in PDFs, this has two downsides I can think of:
- every season we would have to recreate the certificates in HTML in order for us to fill them out (maybe it wouldn't be too bad/too much work)
- it uses `wkhtmltopdf` under the hood which is in archive mode/no longer being maintained (also putting this on our radar to potentially look at alternatives):
  - https://github.com/wkhtmltopdf/wkhtmltopdf
  - https://github.com/wkhtmltopdf/wkhtmltopdf/issues/5160

Anywho, I think there are two potential things for us to explore: HexaPDF and seeing if our existing library will work (w/ a UTF-8 font).

**Comment by shaunxp20** [comment]
I started exploring HexaPDF. I was able to fill in one of our PDFs (a judge certificate with some test/sample data). I then tried to use a special character and it didn't work:
`ruby/3.1.2/lib/ruby/gems/3.1.0/gems/hexapdf-0.32.0/lib/hexapdf/font/type1_wrapper.rb:192:in 'encode': Glyph for "Д" missing (HexaPDF::Error)`

Which I think is because the font being used in the PDF doesn't have support for glyphs/special characters. Which is inline with the message from the current gem (`pdf-forms`) we're using:

> ...your input PDF lacks the proper font for your kind of characters. Re-create it and embed any necessary fonts.

which I think means we need to recreate the PDFs and embed the appropriate fonts.

So, I think that's the next step (maybe for a new ticket), to update the PDFs with different fonts.
